### PR TITLE
stats: new thread local store implementation

### PR DESCRIPTION
**Pull request by mattklein123**
This is needed to support dynamic add/remove of clusters and
transitively stat scopes.

There is no way around the fact that this is an extremely complicated
change. See the comment in thread_local_store.h for more info.

NOTE: This changes the shared memory layout and will require a full
      restart.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team this is a very risky change. I don't plan on shipping till next week and will do some smoke testing so there is time to do a thorough review.

**Comment by mattklein123** [comment]
quick note: I think shutdown is not quite correct. There is a way that we will run runOnAllThreads() while/after workers have shutdown. Will look at this more tomorrow.

**Comment by rshriram** [comment]
@mattklein123 what do you mean by requiring full restart? Are you referring to your existing envoy deployments?

**Comment by mattklein123** [comment]
@rshriram this PR changes the shared memory layout so that the stats can be reference counted. Hot restart is not possible and a full restart is required for deployment. This is not a common occurrence. I think the last time this happened was about a year ago.

**Comment by mattklein123** [comment]
@lyft/network-team I'm smoke testing this now and it looks good so far so I think ready for review.

File: source/common/stats/thread_local_store.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ *   shared across all worker threads.
+ * - Per thread caches are checked, and if empty, they are populated from the central cache.
+ * - Scopes are entirely owned by the caller. The store only keeps weak pointers.
+ * - When a scope is destroyed, a cache flush operation is run on all threads to flush any cached
+ *   data owned by the destroyed scope.
+ * - NOTE: It is theoretically possible that when a scope is deleted, it could be reallocated
+ *         with the same address, and a cache flush operation could race and delete cache data
+ *         for the new scope. This is extremely unlikely, and if it happens the cache will be
+ *         repopulated on the next access.
+ * - Since it's possibly to have overlapping scopes, we de-dup stats when counters() or gauges()
```
Comment (ccaraman) [inline source/common/stats/thread_local_store.h]: typo:possible

**Pull request merged**